// nuxt-svgo/dist/module.mjs
import { defineNuxtModule, createResolver, addVitePlugin, addComponent, extendWebpackConfig } from '@nuxt/kit';
import svgLoader from 'vite-svg-loader';

const nuxtSvgo = defineNuxtModule({
  meta: {
    name: "nuxt-svgo",
    configKey: "svgo",
    compatibility: {
      // Add -rc.0 due to issue described in https://github.com/nuxt/framework/issues/6699
      nuxt: "^3.0.0-rc.0"
    }
  },
  defaults: {
    svgo: true,
    defaultImport: "component",
    svgoConfig: {}
  },
  setup(options) {
    const { resolve } = createResolver(import.meta.url);
    addVitePlugin(svgLoader(options));
    addComponent({
      name: "nuxt-icon",
      global: true,
      filePath: resolve("./runtime/components/nuxt-icon.vue")
    });
    extendWebpackConfig((config) => {
      const svgRule = config.module.rules.find((rule) => rule.test.test(".svg"));
      svgRule.test = /\.(png|jpe?g|gif|webp)$/;
      config.module.rules.push({
        test: /\.svg$/,
        use: [
          "vue-loader",
          {
            loader: "vue-svg-loader",
            options: {
              svgo: false
            }
          },
          options.svgo && {
            loader: "svgo-loader",
            options: options.svgoConfig || {}
          }
        ].filter(Boolean)
      });
    });
  }
});

export { nuxtSvgo as default };
