// nuxt-csurf/dist/module.mjs
import { defineNuxtModule, createResolver, addServerHandler, addServerPlugin, addImports, addPlugin } from '@nuxt/kit';
import { createDefu } from 'defu';

const defuReplaceArray = createDefu((obj, key, value) => {
  if (Array.isArray(obj[key]) || Array.isArray(value)) {
    obj[key] = value;
    return true;
  }
});

const module = defineNuxtModule({
  meta: {
    name: "nuxt-csurf",
    configKey: "csurf"
  },
  defaults: {
    https: process.env.NODE_ENV === "production",
    cookieKey: "",
    cookie: {
      path: "/",
      httpOnly: true,
      sameSite: "strict"
    },
    methodsToProtect: ["POST", "PUT", "PATCH"]
  },
  setup(options, nuxt) {
    const { resolve } = createResolver(import.meta.url);
    if (!options.cookieKey) {
      options.cookieKey = `${options.https ? "__Host-" : ""}csrf`;
    }
    options.cookie = options.cookie || {};
    if (options.cookie.secure === void 0) {
      options.cookie.secure = !!options.https;
    }
    nuxt.options.runtimeConfig.csurf = defuReplaceArray(nuxt.options.runtimeConfig.csurf, { ...options });
    if (options.enabled !== false) {
      addServerHandler({ handler: resolve("runtime/server/middleware/csrf") });
      addServerPlugin(resolve("runtime/server/plugin/csrf"));
    }
    nuxt.options.build.transpile.push(resolve("runtime"));
    addImports(["useCsrf", "useCsrfFetch"].map((key) => ({
      name: key,
      as: key,
      from: resolve("runtime/composables")
    })));
    addPlugin(resolve("runtime/plugin"));
  }
});

export { module as default };
